### ISSUE TITLE
'contains()' & 'in()' filter not working inside 'or()' filter

### ISSUE THREAD

**Issue by godwin5** [supabase/supabase#3444]
![IMG_20211007_184804_154~2.png](https://user-images.githubusercontent.com/28286982/136392857-2dc9608d-1506-4c5f-a933-98e3c6ca4624.png)![IMG_20211007_184807_404~2.png](https://user-images.githubusercontent.com/28286982/136392872-bbff5e25-d7dd-41a5-a5b1-319af7e10500.png)

**Comment by kiwicopple** [comment]
If you're using the `or` function then you need to use the native PostgREST filters: 
https://postgrest.org/en/v8.0/api.html#horizontal-filtering-rows

so this should probably be something like:

```js
supabase
  .from('matches')
  .select()
  .or(`
    players.cs.{p-yinezzhfue}, 
    players.cs.{p-yinezzhfue} 
  `)
```

(note, not tested so you might need to modify it a bit after reading the pgrst docs)

**Comment by godwin5** [comment]
> If you're using the `or` function then you need to use the native PostgREST filters: 
> https://postgrest.org/en/v8.0/api.html#horizontal-filtering-rows
> 
> so this should probably be something like:
> 
> ```js
> supabase
>   .from('matches')
>   .select()
>   .or(`
>     players.cs.{p-yinezzhfue}, 
>     players.cs.{p-yinezzhfue} 
>   `)
> ```
> 
> (note, not tested so you might need to modify it a bit after reading the pgrst docs)

Thanks for this, I'll check

**Comment by GaryAustin1** [comment]
@kiwicopple 
I just spent some time working in Discord with a user who spent awhile themselves on .not with a contains filter.
Postgrest says it should work, so thought it was worth looking at.  Could not find any documentation or issues with .not and .contains in searches on supabase, including here before I started looking at URLS an postgrest needs.

After much playing around I figured out the magic I think...

.not('tags','cs','{a,b}')

Besides guessing to use 'cs' instead of contains (which is confirmed in this report), then had to figure out how to convert the array format .contains uses ['a','b'] to '{a,b}'.

I was about to generate a bug/document report and then searched on filter contains and found this, which sort of addresses part of it, but handled a bit differently for .not in supabase with 3 values needed.

Anyway, this really should be documented (or at least a warning on .or/.and/.not).  Should I generate a documentation issue on this?   Alot of people using supabase.js don't even know anything about postgrest being underneath and that it formats things like arrays and contains differently.

**Comment by kiwicopple** [comment]
>  Should I generate a documentation issue on this? Alot of people using supabase.js don't even know anything about postgrest being underneath and that it formats things like arrays and contains differently.

Yes please Gary - I think you're right, we need to document this one better

**Comment by GaryAustin1** [comment]
@steve-chavez , @kiwicopple 
Adding info to .not, .or, and .filter.

Sorry for taking up time, but I'm giving a shot at actually doing the real document update process with a fork and pull request versus just an issue for someone to do the hard work...

I don't find any real standards for the blocks I'm adding to .not, .or and .filter.  Also adding PostgREST note to .or and .not

Comments before I finish .not and I try and get these changes off of my IDE, back to github and into a pull request?

Samples:
![image](https://user-images.githubusercontent.com/54564956/157777727-750bc2b0-285d-4e64-99d8-b7369163e700.png)
![image](https://user-images.githubusercontent.com/54564956/157777781-f2a056c7-421e-4ec8-8824-76866a2265db.png)

Also does someone maintain the Dart page as should I try and update that too?

**Comment by kiwicopple** [comment]
This seems great @GaryAustin1 - I can come in later and tidy them up

> Also does someone maintain the Dart page or should I try and update that too?

Could you please give this one a go too? Would be great to keep these docs aligned! 🙏

**Comment by steve-chavez** [comment]
Awesome work @GaryAustin1 :1st_place_medal:

**Comment by GaryAustin1** [comment]
@kiwicopple @steve-chavez 
I'm posting this here, but happy to work on a real issue some where...
THINGS ARE  A MESS with special filters beyond just the .not, .or, .filter docs.

### First .contains, .containedBy, .overlap don't really handle ranges and arrays correctly or the same between clients.  

In updating Dart docs and testing I found they wrap the array variable with '( )' for .contains, .containedBy and .overlaps   They also process arrays as ranges in other cases like .not and .filter so don't work on array columns.  The docs for them show an example of an array column with an array search (like a tag array search), but they send a range search.

The JS code wraps the array variable with '{ }' and shows an example of tag array so works well for majority, but won't work for range.

The docs don't mention any thing about ranges/arrays.  I don't know what other clients are doing...

### 2nd Range only operators:

JS Docs show:
![image](https://user-images.githubusercontent.com/54564956/157999155-e07fd285-177b-4597-8dfa-732b062ea7ed.png)
But that errors back from PostgREST as they just pass the array values on with wire..... nxl.150,250 ....
If you put the array in a formatted string with range syntax it is fine. So at minimum Doc issue, but why send just the array values out to begin with...

Dart docs show using a string only!, but seem to also handle an array variable and use correct range syntax on it.

### So Thoughts

At a minimum my personal belief is it is very important to make sure array (tag array type operations) work first for tags as the "easy" case as the use of Postgres range columns have to be much less in your user base.

All clients should react the same.

It might be better to only deal with string value (let user embed array variable in a string) as the way to deal with some of this.

It almost seems like having all these esoteric range options, especially with array options (I'm closing my eyes to what json options are that I see in the code) causes a much larger doc/code/support burden than just saying "Use filter if you are good enough to use ranges :)"

### Conclusion

Hopefully I'm totally off base and missing something...

I've added range examples to the .or, .filter, .not docs I'm working on and will upload as a PR this weekend.

I think there are other things that need looking at before trying to fix the array/range operator docs (although the JS ranges doc error is pretty glaring, I guess must not be used.... much).

**Comment by GaryAustin1** [comment]
@steve-chavez @kiwicopple 
I've generated a 2nd PR with doc changes to .contains, .containedBy and .overlap showing use in both array and range situations based on the current working of postgrest.js.  I also fixed the docs on range only filters so they don't generate errors calling the server by showing using strings for now.  

Still think the whole area of array, range, and native filters needs some love at the actual code side and simpler docs as discussed above, but the new PR helps.

**Comment by steve-chavez** [comment]
Another piece of awesome work @GaryAustin1.

> Still think the whole area of array, range, and native filter formats needs some love at the actual code side

That's true. You know, range types also don't get translated well to json, you need to cast them like in [this example](https://postgrest.org/en/stable/how-tos/casting-type-to-custom-json.html).

That makes me think that instead of using the pg string representation for ranges maybe we should use a json object for filtering them. So something like:

```js
.containedBy('rangecol', { start: 1, end: 20 })
```

WDYT?

**Comment by GaryAustin1** [comment]
If inclusive and exclusive are covered with "flags" that seems clearer to those of who have not encountered range syntax [],[),(],() since college 40 years ago...

Seems like a heavy lift though would it not impact all the PostgREST "drivers" and PostgREST itself?

Right now for only contains and containedBy the JS driver looks for string, then array and then this happens:
      // json
      this.url.searchParams.append(`${column}`, `cs.${JSON.stringify(value)}`)
I have no idea what that can be used successfully for currently.
For your example it puts this on the wire:    cd.{"start":1,"end":20}  though.

**Comment by steve-chavez** [comment]
@GaryAustin1 Oh, I meant that `postgrest-js` would make the convertion from `{ start: 1, end: 20 }` to `(1,20]` internally.

> If you have to know the syntax to set up the range column, then not sure how it helps on the filter side to have another way.

That is true, it wouldn't help in that case.
